### DISCUSSION TITLE
Rx.Subject.never() and Rx.Subject.empty() in rxjs7

### DISCUSSION THREAD

**Discussion by milahu** [ReactiveX/rxjs#6916] [Report issues other than bug]
static methods like `never` or `empty` are deprecated/removed in rxjs7

`Rx.NEVER` and `Rx.EMPTY` are typeof `Observable`

how should i create `empty Subject` and `never Subject`?

```ts
// index.ts
// rxjs version 7.5.5
var enableSubject = false;

var someSubject = enableSubject
  ? new Rx.Subject<SomeType>()
  : Rx.Subject.never() as Rx.Subject<SomeType>; // `Rx.Subject.never()` is rxjs6 api
// typescript error: Property 'never' does not exist on type 'typeof Subject'
// runtime error: TypeError: Rx.Subject.never is not a function

var someSubject = enableSubject
  ? new Rx.Subject<SomeType>()
  : Rx.EMPTY as Rx.Subject<SomeType>;
// typescript error: Conversion of type 'Observable<never>' to type 'Subject<SomeType>'
// may be a mistake because neither type sufficiently overlaps with the other.
// If this was intentional, convert the expression to 'unknown' first.

var someSubject = enableSubject
  ? new Rx.Subject<SomeType>()
  : Rx.EMPTY as unknown as Rx.Subject<SomeType>;
// runtime error: someSubject.next is not a function

var someSubject = new Rx.Subject<SomeType>();
if (!enableSubject) {
  someSubject.next = () => null; // monkey-patch to create a "never subject"
}
// this works, but its dirty
```

api docs

* https://rxjs.dev/api/index/const/NEVER
* https://rxjs.dev/api/index/const/EMPTY
* https://rxjs.dev/api/index/class/Subject
* https://rxjs.dev/api/index/class/Observable

stackoverflow

* [RxJS5 - dummy observables](https://stackoverflow.com/questions/41453418/rxjs5-dummy-observables)

real code

* [@pnpm/default-reporter](https://github.com/pnpm/pnpm/pull/4507)

**Comment by voliva** [comment]
Not sure why would you need it....

But an empty subject is just

```ts
const subject = new Subject();
subject.complete();
// subject will now immediately complete when a new subscriber joins
```

And a "never" subject is just
```ts
const subject = new Subject();
// subject will never emit as long as you don't call `.next()` on it.
```

The concept of empty and never only applies for Observables. It doesn't really make sense to have a "never" subject, as a subject is both an observer and an observable.

**Reply by milahu** [reply]
```ts
const subject = new Subject();
subject.complete();
```

that works : )

documentation is sparse ... https://rxjs.dev/api/index/class/Subject#complete-
but there is https://rxjs.dev/guide/glossary-and-semantics#complete

> Not sure why would you need it....

use case: logging with loglevels. depending on the current loglevel, i want to enable or disable the `Subject`

> subject will never emit as long as you don't call `.next()` on it

that's the point ... i need a subject that emits nothing when i call `subject.next()`
